### PR TITLE
Add global UI

### PR DESCRIPTION
**Pull request by Kiuh**

### CODE REVIEW

File: Tank Survivors/Assets/Scripts/UiPanels/Death.cs [hunk 1]
```diff
@@ -9,14 +14,45 @@ public class Death : MonoBehaviour
...
+        private string repeatSceneName;
+
+        [SerializeField]
+        private string leaveSceneName;
+
         private void Awake()
         {
             tank.OnDeath += ShowLosePanel;
+            repeatButton.onClick.AddListener(() => SceneManager.LoadScene(repeatSceneName));
+            leaveButton.onClick.AddListener(() => SceneManager.LoadScene(leaveSceneName));
```
Comment (morzyolo) [inline Tank Survivors/Assets/Scripts/UiPanels/Death.cs:42]: Why does UI control the transition to another scene?

File: Tank Survivors/Assets/Scripts/Tank/UpgradablePiece/IUpgradablePiece.cs [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    public interface ILeveledUpgrade
+    {
+        public string Guid { get; }
+        public string Description { get; }
+        public uint UpgradingLevel { get; }
+
+        public void ApplyUpgrade(TankImpl tank);
+        public UpgradeVariantInformation GetUpgradeInformation()
+        {
+            return new UpgradeVariantInformation(Description, Guid);
```
Comment (morzyolo) [inline Tank Survivors/Assets/Scripts/Tank/UpgradablePiece/IUpgradablePiece.cs:35]: Why are you creating another variable of value type?

File: Tank Survivors/Assets/Scripts/Tank/Upgrades/SpeedUpgrade.cs [hunk 3]
```diff
@@ -2,50 +2,38 @@
...
         private uint currentLevel = 0;
+        public uint CurrentLevel
+        {
+            get => currentLevel;
+            set => currentLevel = value;
+        }
 
         [SerializeField]
         private uint maxLevel;
+        public uint MaxLevel => maxLevel;
```
Comment (morzyolo) [inline Tank Survivors/Assets/Scripts/Tank/Upgrades/SpeedUpgrade.cs:27]: Why wasn't this implemented from ILeveledSpeedUpgrade if there is a level field here?

File: Tank Survivors/Assets/Scripts/Common/RandomTools.cs [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+using System.Linq;
+using UnityEngine;
+
+namespace Common
+{
+    public static class RandomTools
+    {
+        public static T GetRandom<T>(this List<T> list)
+        {
+            return list[Random.Range(0, list.Count())];
```
[the remark this answers is not in the document]
Comment (Kiuh) [reply Tank Survivors/Assets/Scripts/Common/RandomTools.cs:11]: what `Enumerable.Count()` method, what `List.Count` they are just reading a value from memory

File: Tank Survivors/Assets/Scripts/Tank/TankView.cs [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+        private RectTransform healthBackground;
+
+        [SerializeField]
+        private RectTransform healthFrontBar;
+
+        [SerializeField]
+        private RectTransform experienceBackground;
+
+        [SerializeField]
+        private RectTransform experienceFrontBar;
```
[the remark this answers is not in the document]
Comment (Kiuh) [reply Tank Survivors/Assets/Scripts/Tank/TankView.cs:21]: Ui slider have different appointment, how i want to implement

File: Tank Survivors/Assets/Scripts/UiPanels/LevelUp.cs [hunk 6]
```diff
@@ -9,15 +14,76 @@ public class LevelUp : MonoBehaviour
...
+            {
+                Destroy(upgradeBlocksViewRoot.GetChild(0).gameObject);
+            }
+            foreach (IUpgradablePiece upgrade in upgrades)
+            {
+                UpgradeBlockView upgradeBlock = Instantiate(
+                    upgradeBlockViewPrefab,
+                    upgradeBlocksViewRoot
+                );
+                upgradeBlock.UpgradeName = upgrade.UpgradeName;
```
[the remark this answers is not in the document]
Comment (Kiuh) [reply Tank Survivors/Assets/Scripts/UiPanels/LevelUp.cs:80]: They value and count can change, presently this kind of optimization will be redundant

**Review by morzyolo — APPROVED** [review]
UI still causes scene switching

**Review by morzyolo — COMMENTED** [review]
Try using the MVP pattern. Get events from View and process them in Presenter instead of Controller

**Pull request merged**